freqtrade/freqtradebot.py

```diff
@@ -79,29 +70,20 @@ def __init__(self, config: Dict[str, Any]) -> None:
             self.config.get('edge', {}).get('enabled', False) else None
 
         self.active_pair_whitelist: List[str] = self.config['exchange']['pair_whitelist']
-        self._init_modules()
-
-        # Tell the systemd that we completed initialization phase
-        if self._sd_notify:
-            logger.debug("sd_notify: READY=1")
-            self._sd_notify.notify("READY=1")
-
-    def _init_modules(self) -> None:
-        """
-        Initializes all modules and updates the config
-        :return: None
-        """
-        # Initialize all modules
 
         persistence.init(self.config)
 
-        # Set initial application state
-        initial_state = self.config.get('initial_state')
+    @property
+    def state(self) -> State:
+        if self._worker is None:
+            raise DependencyException("No Worker is available")
+        return self._worker.state
 
-        if initial_state:
-            self.state = State[initial_state.upper()]
-        else:
-            self.state = State.STOPPED
+    @state.setter
+    def state(self, value: State):
+        if self._worker is None:
+            raise DependencyException("No Worker is available")
```

Comment:
Can there be no worker available ...?

anyway, i think we should move state here and remove this "upwards" dependency...

Reply:
> Can there be no worker available ...?

Yes, in the tests that do not change the state of the bot and do not verify the state -- the worker object is not created and thus this variable is None. These are the tests where the constructor of the FreqtradeBot object or get_patched_freqtradebot() are used, without creation of a Worker instance... This is mostly to simplify changes to the tests since in the application itself the Worker instance is definitely created (hmm, what about edge and backtesting modes???).

> anyway, i think we should move state here and remove this "upwards" dependency...

Need to think about this. The state from my point of view is definetely an attribute of the worker. However, moving it back to FreqtradeBot simplifies dependencies alot, right... Hmm... I don't like the bot state to be a member of the FreqtradeBot (it's not trading business logic), but...